src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/RegexNodeConverter.cs

```diff
@@ -40,16 +36,21 @@ internal SymbolicRegexNode<BDD> ConvertToSymbolicRegexNode(RegexNode root)
         {
             Debug.Assert(_builder is not null);
 
-            Stack<(RegexNode Node, bool TryToMarkFixedLength, ConversionResult Result, ConversionResult[]? ChildResults)> work = new();
-            ConversionResult rootres = new();
-            work.Push((root, true, rootres, MkChildResultArray(root.ChildCount())));
+            // Create the root list that will store the built-up result.
+            DoublyLinkedList<SymbolicRegexNode<BDD>> rootResult = new();
 
-            while (work.Count > 0)
+            // Create a stack to be processed in order to process iteratively rather than recursively, and push the root on.
+            Stack<(RegexNode Node, bool TryToMarkFixedLength, DoublyLinkedList<SymbolicRegexNode<BDD>> Result, DoublyLinkedList<SymbolicRegexNode<BDD>>[]? ChildResults)> stack = new();
+            stack.Push((root, true, rootResult, CreateChildResultArray(root.ChildCount())));
+
+            // Continue to iterate until the stack is empty, popping the next item on each iteration.
+            // Some popped items may be pushed back on as part of processing.
+            while (stack.TryPop(out (RegexNode Node, bool TryToMarkFixedLength, DoublyLinkedList<SymbolicRegexNode<BDD>> Result, DoublyLinkedList<SymbolicRegexNode<BDD>>[]? ChildResults) popped))
```

Comment:
Every case except for one was popping after the peek.  I inverted it to always pop and just push the popped item back in the one case where that's needed.